### ISSUE TITLE
Node skips uptime reporting when it is actually connected to tfchain

### ISSUE THREAD

**Issue by scottyeager** [threefoldtech/zos#2377]
A farmer reported an unusual farmerbot violation on node 6694. While investigating, I found that an uptime report was not submitted during the node's wake up.

Here is a sampling of relevant log entries:

```console
[+] powerd: 2024-07-16T07:52:21Z info setting node power state state=true
[+] powerd: 2024-07-16T07:52:20Z info listening for power events
[+] powerd: 2024-07-16T07:52:20Z info enabling wol on interface nic=enp1s0f0
[+] powerd: 2024-07-16T07:52:20Z info node uptime hash hash=0x0000000000000000000000000000000000000000000000000000000000000000
[+] powerd: 2024-07-16T07:52:20Z error node is not healthy skipping uptime reports
[+] powerd: 2024-07-16T07:52:20Z error node can not reach grid services
[+] powerd: 2024-07-16T07:52:20Z info node address address=5H4TnXjryWu8v7gNxTCHM3kwbbqDMf8KMM41XaCV5FprQWet
[+] noded: 2024-07-16T07:52:19Z info setting node public config
[+] noded: 2024-07-16T07:52:19Z info node address address=5H4TnXjryWu8v7gNxTCHM3kwbbqDMf8KMM41XaCV5FprQWet
[+] noded: 2024-07-16T07:52:19Z info node has been registered twin=12216
[-] noded: 2024/07/16 07:52:19 Connecting to wss://03.tfchain.grid.tf/...
```

The first troubling thing here is that Zos has reported that the node is not healthy and cannot reach grid services when it is reading and writing from and to tfchain in the seconds immediately proceeding and following this!

I am also seeing a zero value hash associated with an uptime report for the first time. I can't tell if this is a normal part of the error situation or a bug where Zos wrongly thinks an uptime report was submitted successfully.

The node does not retry sending this uptime report, then later its power target is set to Down by the farmerbot. The node sets its power state to Down then sends a final uptime report and shuts down. From the perspective of minting, this node has only woken up when it is finally shutting down (when the successful uptime report is registered). This confuses the normal flow of the minting process and eventually results in a violation for the node on a later boot, even though the node is booting up successfully when requested.

Bottom line here is that a node *must* send an uptime report before it sets its power state to Down and powers off. If it can't do so immediately, it should retry until successful.

Here's the full sequence of events:

1. Node is standby, power target and state are both Down
2. Power target is changed to Up, node receives WoL packet and powers on
3. **Node submits an uptime report after booting to a sufficient state (this is the part that failed in this example and can't be skipped)**
4. Power target is changed to Down
5. Node changes its power state to Down and powers off **(don't proceed to this step if step 3 failed, wait and retry)**
6. Node sends final uptime report after changing power state to Down

**Comment by scottyeager** [comment]
Here's another instance from mainnet. Two on the same node (6742):

```console
[+] gateway: 2024-07-16T08:21:43Z info starting gateway module broker=unix:///var/run/redis.sock worker nr=1
[+] gateway: 2024-07-16T08:21:43Z error gateway is not supported error="gateway is not supported on this node: no public configuration"
[+] powerd: 2024-07-16T08:21:42Z info setting node power state state=true
[+] powerd: 2024-07-16T08:21:42Z info listening for power events
[+] powerd: 2024-07-16T08:21:42Z info enabling wol on interface nic=enp1s0f0
[+] powerd: 2024-07-16T08:21:42Z info node uptime hash hash=0x0000000000000000000000000000000000000000000000000000000000000000
[+] powerd: 2024-07-16T08:21:42Z error node is not healthy skipping uptime reports
[+] powerd: 2024-07-16T08:21:42Z error node can not reach grid services
[+] powerd: 2024-07-16T08:21:42Z info node address address=5FeAsjgHVFFcYHa4njfb5jkpMDdM2L5YinHYuPVeW6JsRx2r
[+] noded: 2024-07-16T08:21:42Z info setting node public config
[+] noded: 2024-07-16T08:21:42Z info node address address=5FeAsjgHVFFcYHa4njfb5jkpMDdM2L5YinHYuPVeW6JsRx2r
[-] noded: 2024/07/16 08:21:42 Connecting to wss://02.tfchain.grid.tf/...
[+] noded: 2024-07-16T08:21:42Z info node has been registered twin=12290
[+] noded: 2024-07-16T08:21:42Z info node registered node=6742 twin=12290
```

```console
[+] noded: 2024-07-16T21:12:19Z info node address address=5FeAsjgHVFFcYHa4njfb5jkpMDdM2L5YinHYuPVeW6JsRx2r
[-] noded: 2024/07/16 21:12:19 Connecting to wss://04.tfchain.grid.tf/...
[+] noded: 2024-07-16T21:12:19Z info node has been registered twin=12290
[+] noded: 2024-07-16T21:12:19Z info node registered node=6742 twin=12290
[+] powerd: 2024-07-16T21:12:18Z info setting node power state state=true
[+] powerd: 2024-07-16T21:12:18Z info listening for power events
[+] powerd: 2024-07-16T21:12:18Z info enabling wol on interface nic=enp1s0f0
[+] powerd: 2024-07-16T21:12:18Z info node uptime hash hash=0x0000000000000000000000000000000000000000000000000000000000000000
[+] powerd: 2024-07-16T21:12:18Z error node is not healthy skipping uptime reports
[+] powerd: 2024-07-16T21:12:18Z error node can not reach grid services
[+] powerd: 2024-07-16T21:12:18Z info node address address=5FeAsjgHVFFcYHa4njfb5jkpMDdM2L5YinHYuPVeW6JsRx2r
```

I hoped that this might be a transient issue related to the mainnet update, but seeing another instance happening later suggests that this might be ongoing new behavior introduced with the update.

**Comment by scottyeager** [comment]
Mainnet node 7028:

```console
[+] noded: 2024-07-16T08:44:02Z info node has been registered twin=12868
[+] noded: 2024-07-16T08:44:02Z info node registered node=7028 twin=12868
[+] powerd: 2024-07-16T08:44:02Z info listening for power events
[+] powerd: 2024-07-16T08:44:02Z info enabling wol on interface nic=enp1s0f0
[+] powerd: 2024-07-16T08:44:02Z info node uptime hash hash=0x0000000000000000000000000000000000000000000000000000000000000000
[+] powerd: 2024-07-16T08:44:02Z error node is not healthy skipping uptime reports
[+] powerd: 2024-07-16T08:44:02Z error node can not reach grid services
[+] powerd: 2024-07-16T08:44:02Z info node address address=5FajiiiaqC9KDXSPeCY5METD7AHbh8uXxSV8ZbKkzNYmzAF4
```

**Comment by xmonader** [comment]
grid services aren't limited to tfchain, but to every service zos uses. It could be the node can't reach 0-hub or graphql or else. if the node can't reach these services or has a broken disk it will be marked as unhealthy and as good as being offline

https://github.com/threefoldtech/zos/blob/928723db899e48231d326be37152d33353ff9779/pkg/power/uptime.go#L117

**Comment by scottyeager** [comment]
That's helpful information @xmonader, but the node's behavior is still incorrect, even if it can reasonably be deemed unhealthy. Unless there's a corresponding change to minting, which I have also suggested in the linked issue, if the node goes standby without submitting an uptime report it will create wrong outcomes at minting time.

With that said, I find it rather suspicious that nodes which have just bootstrapped themselves from the flist hub not even five minutes ago and have an active connection to tfchain are being marked unhealthy with respect to reaching grid services.

**Comment by scottyeager** [comment]
After some further investigation, I think the source of this particular issue is that the node simply doesn't wait for its network to full come up before concluding that it's unhealthy and skips uptime reporting.

Here's an example to demonstrate, this time for node 6589. The node comes up around 18 Jul 2024 08:16:46 UTC and leases an IP address for `zos`:

```console
2024-07-18 08:16:46	
[-] dhcp-zos: zos: no IPv6 Routers available
2024-07-18 08:16:46	
[-] dhcp-zos: zos: adding default route via 192.168.6.1
2024-07-18 08:16:46	
[-] dhcp-zos: zos: adding route to 192.168.4.0/22
2024-07-18 08:16:46	
[-] dhcp-zos: zos: leased 192.168.5.228 for 21600 seconds
2024-07-18 08:16:46	
[-] dhcp-zos: zos: probing address 192.168.5.228/22
2024-07-18 08:16:46	
[-] dhcp-zos: zos: offered 192.168.5.228 from 192.168.6.1
2024-07-18 08:16:46	
[-] dhcp-zos: zos: soliciting a DHCP lease
2024-07-18 08:16:46	
[-] dhcp-zos: zos: soliciting an IPv6 router
```

A bit later, Zos requests a second lease for `npub4`. One is offered, but not yet leased at this point:

```console
2024-07-18 08:19:50	
[+] networkd: 2024-07-18T08:19:50Z info starting networkd module broker=unix:///var/run/redis.sock worker nr=1
2024-07-18 08:19:50	
[+] networkd: 2024-07-18T08:19:50Z info start zbus server
2024-07-18 08:19:50	
[+] networkd: 2024-07-18T08:19:50Z info filtering out peers from ranges: [{fd00::1 ffffffffffffffff0000000000000000} {305:dc0c:7735:ca69::1 ffffffffffffffff0000000000000000} {205:dc0c:7735:ca69:114a:8d24:84b7:d3b7 fe000000000000000000000000000000}]
2024-07-18 08:19:49	
[+] yggdrasil: 2024/07/18 08:19:49 Your IPv6 subnet is 305:dc0c:7735:ca69::/64
2024-07-18 08:19:49	
[+] yggdrasil: 2024/07/18 08:19:49 Your IPv6 address is 205:dc0c:7735:ca69:114a:8d24:84b7:d3b7
2024-07-18 08:19:49	
[+] yggdrasil: 2024/07/18 08:19:49 Your public key is 048fce2328d65bbad5cb6ded20b120838abddd139f9e778ca58c3aa1dfe48814
2024-07-18 08:19:49	
[+] yggdrasil/test: 2024/07/18 08:19:49 Fatal error: dial unix /var/run/yggdrasil/yggdrasil.sock: connect: no such file or directory
2024-07-18 08:19:49	
[+] yggdrasil/test: 2024/07/18 08:19:49 Connecting to UNIX socket /var/run/yggdrasil/yggdrasil.sock
2024-07-18 08:19:49	
[+] yggdrasil/test: 2024/07/18 08:19:49 Falling back to platform default unix:///var/run/yggdrasil/yggdrasil.sock
2024-07-18 08:19:49	
[+] yggdrasil/test: 2024/07/18 08:19:49 Can't open config file from default location /etc/yggdrasil/yggdrasil.conf
2024-07-18 08:19:48	
[+] networkd: 2024-07-18T08:19:48Z error failed to lookup ip error="lookup jorropo.net on 192.168.6.1:53: server misbehaving" url=tls://jorropo.net:12221?key=060f2d49c6a1a2066357ea06e58f5cff8c76a5c0cc513ceb2dab75c900fe183b&sni=jorropo.net
2024-07-18 08:19:48	
[+] networkd/test: ## Status for  network
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: probing address 192.168.6.226/22
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: offered 192.168.6.226 from 192.168.6.1
2024-07-18 08:19:46	
[+] networkd/test: 2024-07-18T08:19:46Z fatal exiting error="context deadline exceeded"
2024-07-18 08:19:46	
[+] networkd/test: 2024-07-18T08:19:46Z error failed to get status for module error="context deadline exceeded" module=network
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: soliciting a DHCP lease
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: soliciting an IPv6 router
2024-07-18 08:19:46	
[-] dhcp-npub4: ipv6_addaddr1: Permission denied
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: adding address fe80::f011:3cbc:6734:9b2b
2024-07-18 08:19:46	
[-] dhcp-npub4: npub4: IAID dd:32:3d:0f
2024-07-18 08:19:46	
[-] dhcp-npub4: DUID 00:04:36:47:30:44:41:38:5a:43:33:36:32:36:43:4c:58:4e
2024-07-18 08:19:46	
[+] networkd: 2024-07-18T08:19:46Z info filtering out peers from ranges: [{fd00::1 ffffffffffffffff0000000000000000}]
2024-07-18 08:19:46	
[+] networkd: 2024-07-18T08:19:46Z info create dhcp-npub4 zinit service
```

Now the `networkd` module is started, and this seems to be the trigger for a number of network related things to start happening. Crucially, within the same second, network health checks have failed:

```console
2024-07-18 08:19:50	
[+] noded: 2024-07-18T08:19:50Z warn registration failed error="fetch location: failed to fetch location information" sleep=311.512232ms
2024-07-18 08:19:50	
[+] noded: 2024-07-18T08:19:50Z warn the node isn't ready yet err="registration is still in progress" sleep=289.171784ms
2024-07-18 08:19:50	
[+] noded: 2024-07-18T08:19:50Z error failed to fetch location from geoip service error="Get \"https://03.geoip.grid.tf/\": dial tcp: lookup 03.geoip.grid.tf on 192.168.6.1:53: dial udp 192.168.6.1:53: connect: network is unreachable" url=https://03.geoip.grid.tf/
2024-07-18 08:19:50	
[+] noded: 2024-07-18T08:19:50Z error failed health check. retrying error="failed health check" check=network retry-in=16913.956828
```

About one second later the node finally completes its DHCP lease for `npub4`:

```console
2024-07-18 08:19:51	
[-] dhcp-npub4: npub4: adding default route via 192.168.6.1
2024-07-18 08:19:51	
[+] noded: 2024-07-18T08:19:51Z error failed to run ntp check error="failed to get date: Get \"https://worldtimeapi.org/api/timezone/UTC\": dial tcp: lookup worldtimeapi.org on 192.168.6.1:53: dial udp 192.168.6.1:53: connect: network is unreachable"
2024-07-18 08:19:51	
[-] dhcp-npub4: npub4: adding route to 192.168.4.0/22
2024-07-18 08:19:51	
[-] dhcp-npub4: npub4: leased 192.168.6.226 for 21600 seconds
```

At this point the network errors start to quiet down, and it looks like this is when Zos is finally able to reach the outside world. I don't fully understand that, because I thought that this traffic would go over `zos` and not `npub4`. Maybe it's just a coincidence.

About ten seconds after that, the node decides not to send an uptime report due to the earlier failed check:

```console
2024-07-18 08:20:01	
[+] powerd: 2024-07-18T08:20:01Z info node uptime hash hash=0x0000000000000000000000000000000000000000000000000000000000000000
2024-07-18 08:20:01	
[+] powerd: 2024-07-18T08:20:01Z error node is not healthy skipping uptime reports
2024-07-18 08:20:01	
[+] powerd: 2024-07-18T08:20:01Z error node can not reach grid services
```

Tragically, that's only about six seconds before the health checks would be retried. Given the absence of any more errors, it looks like the health checks indeed passed on the second try but the uptime reporting wasn't retried.

**Comment by ramezsaeed** [comment]
Fixed in :
https://github.com/threefoldtech/zos/issues/2381
